dask/array/linalg.py

```diff
@@ -201,6 +201,96 @@ def tsqr(data, name=None, compute_svd=False):
         return u, s, v
 
 
+def sfqr(data, name=None):
+    """ Direct Short-and-Fat QR
+
+    Currently, this is a quick hack for non-tall-and-skinny matrices which
+    are one chunk tall and (unless they are one chunk wide) have chunks
+    that are wider than they are tall
+
+    Q [R_1 R_2 ...] = [A_1 A_2 ...]
+
+    it computes the factorization Q R_1 = A_1, then computes the other
+    R_k's in parallel.
+
+    Parameters
+    ----------
+
+    data: Array
+
+    See Also
+    --------
+
+    dask.array.linalg.qr - Supported by this
+    """
+
+    nr, nc = len(data.chunks[0]), len(data.chunks[1])
+    cr, cc = data.chunks[0][0], data.chunks[1][0]
+
+    if not ((data.ndim == 2) and  # Is a matrix
+            (nr == 1) and         # Has exactly one block row
+            ((cr <= cc) or        # Chunking dimension on rows is at least that on cols or...
+             (nc == 1))):         # ... only one block col
+        raise ValueError(
+            "[sfqr] Input must have the following properties:\n"
+            "  1. Have two dimensions\n"
+            "  2. Have only one row of blocks\n"
+            "  3. Either one column of blocks or chunking dimension on rows is at most that on cols"
+        )
```

Comment:
We should possibly avoid using terms like sfqr in user-facing error messages.  Users just calling `da.linalg.qr` might not understand this acronym.  

We might also consider moving these informative error messages up to `da.linalg.qr`

Reply:
Will keep a reference to sfqr, but also refer to qr and the relevant case.
```
    if not ((data.ndim == 2) and  # Is a matrix
            (nr == 1) and         # Has exactly one block row
            ((cr <= cc) or        # Chunking dimension on rows is at least that on cols or...
             (nc == 1)) and       # ... only one block col
            _has_even_chunks(data.chunks[1])  # Has uniform chunking on column dimension
           ):
        raise ValueError(
            "Input must have the following properties:\n"
            "  1. Have two dimensions\n"
            "  2. Have only one row of blocks\n"
            "  3. Either one column of blocks or chunking dimension on rows\n"
            "     is at most that on cols (for the first block, if uneven)\n"
            "  4. Has even chunking on column dimension (e.g.: (5,5,5) or\n"
            "     (4,4,3); the last column is the only one allowed to differ)\n\n"
            "Note: This function (sfqr) supports QR decomposition in the case\n"
            "of short-and-fat matrices (single row chunk/block; see qr)"
        )
```

Some duplication in `qr`:
```
    if len(a.chunks[1]) == 1:
        return tsqr(a, name)
    elif len(a.chunks[0]) == 1:
        return sfqr(a, name)
    else:
        raise NotImplemented(
            "qr currently supports only tall-and-skinny (single column chunk/block; see tsqr)\n"
            "and short-and-fat (single row chunk/block; see sfqr) matrices\n\n"
            "Consider use of the rechunk method. For example,\n\n"
            "x.rechunk({0: -1, 1: 'auto'}) or x.rechunk({0: 'auto', 1: -1})\n\n"
            "which rechunk one shorter axis to a single chunk, while allowing\n"
            "the other axis to automatically grow/shrink appropriately."
        )
```
... but ok because "pure functions".

Reply:
... will get rid of requirement 4 which I added. Will support non-uniform chunking like
```
(
  (10),
  (15, 8, 7, 5, 5)
)
```